awx/main/scheduler/task_manager_models.py

```diff
@@ -0,0 +1,120 @@
+                hostname=instance.hostname,
+            )
+            for instance in instances
+        ]
+        self._update_remaining_capacity(active_tasks)
+        self.instances_by_hostname = {i.hostname: i for i in instance_list}
+
+    def _update_remaining_capacity(self, tasks):
+        """Takes a list of tasks and updates the remaining capacity on our TaskManagerInstance list.
+
+        Computes remaining capacity for all the instances based on currently running and waiting tasks.
+        """
+        if self._remaining_capacity_updated:
+            # Only want to do this once
+            return
+        for task in tasks:
+            if task.status not in ['waiting', 'running']:
+                continue
+            control_instance = self.instances_by_hostname.get(task.controller_node, '')
+            execution_instance = self.instances_by_hostname.get(task.execution_node, '')
+            if execution_instance and execution_instance.node_type in ('hybrid', 'execution'):
+                self.instances_by_hostname[task.execution_node].remaining_capacity -= task.task_impact
+            if control_instance and control_instance.node_type in ('hybrid', 'control'):
+                self.instances_by_hostname[task.controller_node].remaining_capacity -= settings.AWX_CONTROL_NODE_TASK_IMPACT
+
+    def __getitem__(self, hostname):
+        return self.instances_by_hostname.get(hostname)
+
+    def __contains__(self, hostname):
+        return hostname in self.instances_by_hostname
+
+
+class TaskManagerInstanceGroups:
+    """A class representing minimal data the task manager needs to represent an InstanceGroup."""
+
+    def __init__(self, instances_by_hostname):
+        self.instance_groups = dict()
+        self.controlplane_ig = None
+
+        for instance_group in InstanceGroup.objects.prefetch_related('instances').only('name', 'instances'):
+            if instance_group.name == settings.DEFAULT_CONTROL_PLANE_QUEUE_NAME:
+                self.controlplane_ig = instance_group
+            self.instance_groups[instance_group.name] = dict(
+                instances=[
+                    instances_by_hostname[instance.hostname] for instance in instance_group.instances.all() if instance.hostname in instances_by_hostname
+                ],
+            )
+
+    def __getitem__(self, instance_group_name):
+        return self.instance_groups.get(instance_group_name)
+
+    def fit_task_to_most_remaining_capacity_instance(
+        self, task, instance_group_name=None, impact=None, capacity_type=None, add_hybrid_control_cost=False, instances=[]
+    ):
+        impact = impact if impact else task.task_impact
+        capacity_type = capacity_type if capacity_type else task.capacity_type
+        instance_most_capacity = None
+        most_remaining_capacity = -1
+        assert instances or instance_group_name, "Need to provide either the name of an instance group or a list of instances"
```

Comment:
I'm a little thrown off by this method signature. It looks like we could get rid of the `instances` keyword argument. Also, other people will jump on your case for using a mutable default of `[]`. This parameter is a liability that we would do better to get rid of.

It seems that the tests use it, and modifying them is challenging without database access. Removing `instances` would require some re-configuration so that we can probably initialize everything in this file with data from outside the database. This is generally a good idea anyway.

To start with `TaskManagerInstances.__init__` will need to optionally accept a _list_ of instances as an alternative to fetching the queryset. That's the right way to do things. I think it's a good idea. Then in the unit test, you construct all of these objects from basic python.

Reply:
Great ideas! Yes this is why I was accepting the list, because of the tests